File: Cpp/AdventOfCode/src/2016/d6_RepeatSignal.cpp
#include "2016/d6_RepeatSignal.h"

SOLUTION(2016, 6) {
    template<size_t MsgLength>
    using CountType = std::array<std::array<u32, 26>, MsgLength>;

    template<size_t MsgLength>
    constexpr void UpdateLetterCounts(std::string_view line, CountType<MsgLength>&result) {
        for (size_t i = 0; i < line.size(); i++) {
            result[i][line[i] - 'a']++;
        }
    }

    template<size_t MsgLength>
    constexpr std::string Decode(CountType<MsgLength>&counts) {
        std::string result = "";
        for (auto i = 0; i < MsgLength; i++) {
            u32 best = 0;
            u32 bestIndex = 0;
            for (auto count = 0; count < 26; count++) {
                if (counts[i][count] > best) {
                    best = counts[i][count];
                    bestIndex = count;
                }
            }

            result += static_cast<char>(bestIndex + 'a');
        }

        return result;
    }

    template<size_t MsgLength>
    constexpr std::string DecodeMin(CountType<MsgLength>&counts) {
        std::string result = "";
        for (auto i = 0; i < MsgLength; i++) {
            u32 best = 999;
            u32 bestIndex = 0;
            for (auto count = 0; count < 26; count++) {
                if (counts[i][count] > 0 && counts[i][count] < best) {
                    best = counts[i][count];
                    bestIndex = count;
                }
            }
            result += static_cast<char>(bestIndex + 'a');
        }

        return result;
    }

    constexpr std::string Solve(auto Decoder) {
        CountType<8> counts{};
        for (auto line : Lines) {
            UpdateLetterCounts(line, counts);
        }
        return Decoder(counts);
    }

    PART_ONE() {
        return Solve(Decode<8>);
    }
    PART_TWO() {
        return Solve(DecodeMin<8>);
    }

    TESTS() {
        CountType<6> counts{};

        UpdateLetterCounts("eedadn", counts);
        UpdateLetterCounts("drvtee", counts);
        UpdateLetterCounts("eandsr", counts);
        UpdateLetterCounts("raavrd", counts);
        UpdateLetterCounts("atevrs", counts);
        UpdateLetterCounts("tsrnev", counts);
        UpdateLetterCounts("sdttsa", counts);
        UpdateLetterCounts("rasrtv", counts);
        UpdateLetterCounts("nssdts", counts);
        UpdateLetterCounts("ntnada", counts);
        UpdateLetterCounts("svetve", counts);
        UpdateLetterCounts("tesnvt", counts);
        UpdateLetterCounts("vntsnd", counts);
        UpdateLetterCounts("vrdear", counts);
        UpdateLetterCounts("dvrsen", counts);
        UpdateLetterCounts("enarar", counts);

        if (counts[0]['e' - 'a'] == 0) return false;

        if (Decode(counts) != "easter") return false;
        if (DecodeMin(counts) != "advent") return false;
        return true;
    }
}

File: Rust/rust_aoc/src/year_2016/d6_repeat_signal.rs
use itertools::Itertools;

pub fn part_one(lines: &Vec<String>) -> String {
    decode(lines, false)
}
pub fn part_two(lines: &Vec<String>) -> String {
    decode(lines, true)
}

fn decode(lines: &Vec<String>, use_min: bool) -> String {
    let first_letters: Vec<String> = (0..lines[0].len())
        .map(|i| lines.iter().map(|line| line.as_bytes()[i] as char).collect())
        .collect();

    first_letters.iter().map(|letters| {
        let letter_counts = letters.chars().into_group_map_by(|&x| x)
        .into_iter().map(|(k, v)| (k, v.len()));
        let (letter, _count) = match use_min {
            true => letter_counts.min_by(|(_, lhs), (_, rhs)| lhs.cmp(rhs)).unwrap(),
            false => letter_counts.max_by(|(_, lhs), (_, rhs)| lhs.cmp(rhs)).unwrap()
        };
        letter
        }
    ).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn first_example() {
        let entries = vec!["eedadn", "drvtee", "eandsr", "raavrd", "atevrs", "tsrnev", "sdttsa", "rasrtv", "nssdts", "ntnada", "svetve", "tesnvt", "vntsnd", "vrdear", "dvrsen", "enarar"].into_iter().map(|s| String::from(s)).collect();
        assert_eq!("easter", decode(&entries, false));
        assert_eq!("advent", decode(&entries, true));

    }
}
